File: 백준/Gold/14890. 경사로/경사로.py
def check_line(line):
    global cnt
    start = line[0]
    now = 1
    flag = False
    for i in range(1,N):
        try:
            if line[i-1] == line[i+1] and line[i-1] > line[i]:
                return
        except:
            pass

        # 2칸 이상 차이나면 바로 끝
        if abs(line[i]-start) > 1:
            # print("2칸 이상 차이")
            return
        # 같은 높이인 경우 경사로 놓일 공간 +
        elif line[i] == start:
            now += 1
            # 만약 경사로를 세우는 상황이였고 경사로가 놓일 자리가 생겼다면 초기화
            if flag and now >= L:
                flag = False
                now = 0

        # 올라가는 경사로가 필요한 경우
        elif line[i] > start:
            # 경사로를 세우는 상황이 아니고 경사로 놓일 자리가 있다면 세우기
            if flag == False and now >= L:
                flag = False
                now = 1
            # 그 외는 안됨
            else:
                # print("올라가다가 오류가 남",i)
                return
            start = line[i]
        # 내려가는 경사로가 필요한 경우
        elif line[i] < start:
            # 이미 경사로 세우고 있었으면 안됨
            if flag:
                # print("내려가다가 오류가 남", i)
                return
            now = 1
            start = line[i]
            if L == 1:
                continue
            flag = True
    if flag:
        return
    # print("통과", line)
    cnt += 1



N, L = map(int,input().split())
arr = []
cnt = 0
for _ in range(N):
    arr.append(list(map(int,input().split())))
for i in range(N):
    vertical_line = []
    horizontal_line = []
    for j in range(N):
        vertical_line.append(arr[i][j])
        horizontal_line.append(arr[j][i])
    # print(vertical_line,horizontal_line)
    check_line(vertical_line)
    check_line(horizontal_line)
print(cnt)

File: 백준/Gold/14890. 경사로/경사로.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int l = sc.nextInt();
        int cnt = 0;
        int[][] arr = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j] = sc.nextInt();
            }
        }
        for (int i = 0; i < n; i++) {
            List<Integer> verticalLine = new ArrayList<>();
            List<Integer> horizontalLine = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                verticalLine.add(arr[i][j]);
                horizontalLine.add(arr[j][i]);
            }

            if (check(verticalLine, l)) {
                cnt += 1;
            }
            if (check(horizontalLine, l)) {
                cnt += 1;
            }
        }
        System.out.println(cnt);
    }

    private static boolean check(List<Integer> arr, int l) {
        int n = arr.size();
        int prev = arr.get(0);
        int curGround = 1;
        boolean isSettled = false;

        for (int i = 1; i < n; i++) {
            if (i < n - 1 && arr.get(i - 1).equals(arr.get(i + 1)) && arr.get(i - 1) > arr.get(i)) {
                return false;
            }

            if (Math.abs(arr.get(i) - prev) > 1) {
                return false;
            } else if (arr.get(i).equals(prev)) {
                curGround += 1;
                if (isSettled && curGround >= l) {
                    isSettled = false;
                    curGround = 0;
                }
            } else if (arr.get(i) > prev) {
                if (!isSettled && curGround >= l) {
                    isSettled = false;
                    curGround = 1;
                } else {
                    return false;
                }
                prev = arr.get(i);
            } else if (arr.get(i) < prev) {
                if (isSettled) {
                    return false;
                }
                curGround = 1;
                prev = arr.get(i);
                if (l == 1) {
                    continue;
                }
                isSettled = true;
            }
        }
        return !isSettled;
    }
}
